src/components/smart_objects/src/object_schema_item.cc

```diff
@@ -99,14 +159,106 @@ Errors::eType CObjectSchemaItem::validate(const SmartObject& object,
   return Errors::OK;
 }
 
-void CObjectSchemaItem::applySchema(SmartObject& Object,
-                                    const bool RemoveFakeParameters) {
+Errors::eType CObjectSchemaItem::validate(
+    const SmartObject& object,
+    rpc::ValidationReport* report__,
+    const utils::SemanticVersion& MessageVersion) {
+  if (SmartType_Map != object.getType()) {
+    std::string validation_info = "Incorrect type, expected: " +
+                                  SmartObject::typeToString(SmartType_Map) +
+                                  ", got: " +
+                                  SmartObject::typeToString(object.getType());
+    report__->set_validation_info(validation_info);
+    return Errors::INVALID_VALUE;
+  }
+
+  std::set<std::string> object_keys = object.enumerate();
+
+  for (Members::const_iterator it = mMembers.begin(); it != mMembers.end();
+       ++it) {
+    const std::string& key = it->first;
+    const SMember& member = it->second;
+    std::set<std::string>::const_iterator key_it = object_keys.find(key);
+    if (object_keys.end() == key_it) {
+      if (member.mSince.is_initialized() &&
+          MessageVersion < member.mSince.get() &&
+          member.mHistoryVector.size() > 0) {
+        // Message version predates parameter and a history vector exists.
+        for (uint i = 0; i < member.mHistoryVector.size(); i++) {
+          if (member.mHistoryVector[i].mSince.is_initialized() &&
+              MessageVersion >= member.mHistoryVector[i].mSince.get()) {
+            if (member.mHistoryVector[i].mUntil.is_initialized() &&
+                MessageVersion >= member.mHistoryVector[i].mUntil.get()) {
+              // MessageVersion is newer than the specified "Until" version
+              continue;
+            } else {
+              if (member.mHistoryVector[i].mIsMandatory == true &&
+                  (member.mHistoryVector[i].mIsRemoved == false)) {
+                std::string validation_info =
```

Comment:
@JackLivio is it possible to refactor so complicated method to multiple methods, maybe lambdas ?